fsm_internal.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
```

Comment:
Haha, fair. They aren't that bad once you remove all the backslashes imo :P
I'm a little torn on the level of documentation needed for the macros. I've already explained the general idea of how they work, and I don't want to end up duplicating the code as a comment.